libevmasm/ThirdOptimizerTesting.cpp

```diff
@@ -0,0 +1,368 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libevmasm/ThirdOptimizerTesting.h>
+#include "ThirdOptimizerTesting.h"
+#include <algorithm>
+#include <tuple>
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+It noReturnValueBegin(It _begin, It _end);
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+It backExpressionBegin(It _begin, It _end, bool _allowNoReturn = false)
+{
+	AssemblyItem const& topLevelItem = *(_end - 1);
+
+	signed arguments = topLevelItem.arguments();
+	auto currentEnd = _end - 1 /* top-level operation */;
+	while (arguments > 0)
+	{
+		// TODO this code might break if anything returns more than 1 value, since we can't split things up
+		currentEnd = noReturnValueBegin(_begin, currentEnd);
+		arguments -= static_cast<AssemblyItem const&>(*(currentEnd - 1)).returnValues();
+		currentEnd = backExpressionBegin(_begin, currentEnd, false);
+		assertThrow(currentEnd >= _begin, OptimizerException, "currentEnd is before begin");
+	}
+
+	if (!_allowNoReturn && topLevelItem.returnValues() == 0) return backExpressionBegin(_begin, currentEnd, false);
+
+	return currentEnd;
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+It noReturnValueBegin(It _begin, It _end)
+{
+	assertThrow(_begin != _end, OptimizerException, "expression is empty");
+	AssemblyItem const& topLevelItem = *(_end - 1);
+	if (topLevelItem.returnValues() != 0) return _end;
+	return backExpressionBegin(_begin, _end, true);
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+vector<It> parseExpressions(It _begin, It _end, signed _limit = -1, bool _allowNoReturn = false)
+{
+	auto currentEnd = _end;
+	std::vector<It> expressions{};
+	while (currentEnd != _begin && (_limit == -1 || signed(expressions.size()) < _limit))
+	{
+		auto oldEnd = currentEnd;
+		currentEnd = backExpressionBegin(_begin, oldEnd, _allowNoReturn);
+		expressions.push_back(currentEnd);
+		assertThrow(currentEnd >= _begin, OptimizerException, "currentEnd is before begin");
+	}
+
+	return expressions;
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+vector<It> parseArguments(It _begin, It _end)
+{
+	auto size = _end - _begin;
+	assertThrow(size > 0, OptimizerException, "cannot get arguments from empty _items vector");
+
+	unsigned neededArguments = (_end - 1)->arguments();
+
+	auto expressions = parseExpressions(_begin, _end - 1, neededArguments, false);
+	assertThrow(expressions.size() >= neededArguments, OptimizerException, "more arguments needed than provided");
+
+	expressions = vector<It>(expressions.begin(), expressions.begin() + neededArguments);
```

Comment:
Why do you need to return a vector instead of a pair of iterators?

Reply:
That would be equally valid, but iterators are (hopefully) cheap to copy. And, an iterator to iterator is just a little bit confusing.

Reply:
Oh, and the vector is dead at that point, so you would have iterators into a destructed vector.